dipy/tracking/utils.py

```diff
@@ -484,6 +489,45 @@ def target(streamlines, target_mask, affine, include=True):
             yield sl
 
 
+def near_roi(streamlines, target_mask, affine=None, tol=0):
+    """
+    Provide filtering criteria for a set of streamlines based on whether they
+    fall within a tolerance distance from an ROI
+
+    Parameters
+    ----------
+    streamlines: list
+        A sequence of streamlines. Each streamline should be a (N, 3) array,
+        where N is the length of the streamline.
+    target_mask: ndarray
+        A mask used as a target. Non-zero values are considered to be within
+        the target region.
+    affine: ndarray
+        Affine transformation from voxels to streamlines. Default: identity.
+    tol: float
+        Distance (in the units of the streamlines, usually mm). If any
+        coordinate in the streamline is within this distance from any voxel in
+        the ROI, the filtering criterion is set to True for this streamlin,
+        otherwise False. Default: 0
+
+    Returns
+    -------
+    Boolean array. Set to True in each streamline
+    """
+    if affine is None:
+        affine = np.eye(4)
+    roi_coords = np.array(np.where(target_mask)).T
+    x_roi_coords = apply_affine(affine, roi_coords)
+    out = np.zeros(len(streamlines), dtype=bool)
+    for ii, sl in enumerate(streamlines):
+        for coord in sl:
```

Comment:
Hi Ariel,

A use case to try is a huge ROI. For example take the fornix bundle (or a
few thousand random streamlines) and perform an operation using a gray
matter mask as an ROI. This is the case that will typically bog down the
computation and would hopefully show a substantial speed up.


Jason D. Yeatman, PhD
Research Scientist
Institute for Learning and Brain Sciences (ILABS)
University of Washington
http://jasonyeatman.com

On Sat, May 2, 2015 at 1:30 PM, Ariel Rokem <notifications@github.com>
wrote:

> In dipy/tracking/utils.py
> <https://github.com/nipy/dipy/pull/635#discussion_r29550437>:
>
> > +        Distance (in the units of the streamlines, usually mm). If any
> > +        coordinate in the streamline is within this distance from any voxel in
> > +        the ROI, the filtering criterion is set to True for this streamlin,
> > +        otherwise False. Default: 0
> > +
> > +    Returns
> > +    -------
> > +    Boolean array. Set to True in each streamline
> > +    """
> > +    if affine is None:
> > +        affine = np.eye(4)
> > +    roi_coords = np.array(np.where(target_mask)).T
> > +    x_roi_coords = apply_affine(affine, roi_coords)
> > +    out = np.zeros(len(streamlines), dtype=bool)
> > +    for ii, sl in enumerate(streamlines):
> > +        for coord in sl:
>
> What I have done so far seems to have resulted in only a very small
> speed-up. I am using the fornix bundle as streamlines:
>
> https://gist.github.com/arokem/51239fe55c07da101633
>
> This level of cythonization takes me from about 730 (only python) to about
> 650 msec (python + cython), so about 10% speedup
>
> Happy to hear what else should be done in the cython code to really take
> advantage of it, and increase the speedup more.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/635/files#r29550437>.
>